### ISSUE TITLE
Trash doesn't work wirth BTRFS subvolumes

### ISSUE THREAD

**Issue by blurayne** [andreafrancia/trash-cli#353]
**Describe the bug**

I regularly do snapshots on my BTRFS filesystem. Anyway, I don't want Trash to be part of it so the easiest way to do this is to create a BTRFS subvolume. 

Anyway, this doesn't work with Trash CLI since "Volume" is now different

```bash
$ trash-put foo.go
trash-put: Volume of file: /home
trash-put: Trash-dir: /home/ctang/.local/share/Trash from volume: /home/ctang/.local/share/Trash
trash-put: found unusable .Trash dir (should be a dir): /home/.Trash
trash-put: Trash-dir: /home/.Trash-1000 from volume: /home
trash-put: Failed to trash shell.go in /home/.Trash-1000, because :[Errno 13] Permission denied: '/home/.Trash-1000'
trash-put: cannot trash regular file 'oo.go'
```

**Are you using the latest version of trash-cli?**

Yes

**To Reproduce**

1. BTRFS filesystem
2. Turn $HOME/.local/share/Trash into a subvolume
3. Ensure permissions and ownership of $HOME/.local/share/Trash stay the same
4. Delete any file on the volume.

**Expected behavior**

App should detect that trash file is a subvolume on the same volume and maybe ask if you want to move file there or automatically be able by providing flag or config option. Maybe make a difference if it is under the current subvolume or apart from it on the same logical volume.

**Volumes detail**

```bash
$ trash-list --debug-volumes    
(empty)
```

_Interesting…?!_

**Comment by andreafrancia** [comment]
trash-cli does not support BRTFS, sorry.

**Comment by andy5995** [comment]
I implemented this in [rmw v0.9.3](https://github.com/theimpossibleastronaut/rmw/releases/tag/v0.9.3)

@andreafrancia I had to do a bit of investigating on how to do this. I found this in GNU coretutils copy.c (which is used by GNU coreutils mv.c):

```
/* Perform the O(1) btrfs clone operation, if possible.
   Upon success, return 0.  Otherwise, return -1 and set errno.  */
static inline int
clone_file (int dest_fd, int src_fd)
{
#ifdef FICLONE
  return ioctl (dest_fd, FICLONE, src_fd);
#else
  (void) dest_fd;
  (void) src_fd;
  errno = ENOTSUP;
  return -1;
#endif
}
```

Most of my implementation is at https://github.com/theimpossibleastronaut/rmw/blob/94bc8b3b06404717021f0d9493a403e1ebf4ba84/src/btrfs.c

Basically a btrfs subvolume has a different device id than the root volume. Usually in C, rename() will do a move without a copy and delete. If it's on a different device, it will fail.

I tried some code to test if a trash folder is a subvolume but I couldn't get it to work without requiring sudo. Which may stand to reason because sudo is needed at the command line to do something simple like:

```
sudo btrfs subvolume list .
```

So what I finally settled on is testing if a filesystem is btrfs, if so, try a clone. If that fails, keep iterating through the trash folders until dev_id matches, or if btrfs, try clone again.

I understand you're using python. ;) My point is, I would suspect there's some python library or function equivalent to the C implementation. As for Mac and BSD, I don't support that yet. From what I could gather from my reading, only Linux natively supports btrfs.

I have a 256MB filesystem img that I use in my CI and locally for testing. You can download it and use for testing if you like:

https://www.dropbox.com/scl/fi/57g3ixd3w3tuz4qoc2zp1/rmw-btrfs-test.img?rlkey=yc7krtntswsa1bwz0sbugy4gi&st=hkgrht05&dl=0

```
sudo mount -o loop "$IMAGE_PATH" \
    "$BTRFS_IMAGE_MOUNTPOINT"
```

Where $BTRFS_IMAGE_MOUNTPOINT can be a directory under /tmp or wherever you like (hint: you can use that subvolume command I mentioned above after it's mounted).

**Comment by andy5995** [comment]
@andreafrancia I forgot to mention, the clone operation is FAST even on a 8GB file (<1s). It's different from a copy and delete. You can try with GNU coreutils mv on the btrfs filesystem image I linked above.

**Comment by jtrees** [comment]
> I implemented this in [rmw v0.9.3](https://github.com/theimpossibleastronaut/rmw/releases/tag/v0.9.3)

Thanks @andy5995. This reason was compelling enough for me to switch from trash-cli to rmw. I like it so far!

**Comment by andreafrancia** [comment]
@blurayne Thank you for the information you provided.
I cant no more work on trash-cli.
I added not the README indicating the presence of rmw as a software that can substitute trash-cli when BRTFS is needed.
